projects/ngx-charts-on-fhir/src/lib/fhir-data/fhir-data.service.ts

```diff
@@ -95,15 +98,64 @@ export class FhirDataService {
       // use maximum page size on every request to improve performance
       url = addCountParam(url);
       const request = currentPatientOnly ? this.client.patient.request.bind(this.client.patient) : this.client.request.bind(this.client);
+
       request(url, { pageLimit: 0, onPage })
         .then(() => subscriber.complete())
         .catch((error) => subscriber.error(error));
       return teardownLogic;
-    });
+    }).pipe(
+      retryBackoff({
+        initialInterval: this.INITIAL_INTERVAL,
+        maxRetries: this.MAX_RETRIES,
+        resetOnSuccess: true,
+        shouldRetry: (error) => {
+          return this.errorStatusCheck(error); // always retry
+        },
+      })
+    );
   }
 
-  addPatientData(resource: Resource) {
-    return this.client?.create(resource);
+  errorStatusCheck(error: any) {
+    switch (error.status) {
+      case 500:
+        return true;
+      case 501:
+        return true;
+      case 502:
+        return true;
+      case 503:
+        return true;
+      case 504:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  addPatientData<T extends Resource>(resource: T): Observable<T> {
+    return new Observable<T>((subscriber) => {
+      if (!this.client) {
+        subscriber.error('FhirClientService has not been initialized.');
+        return;
+      }
+      const request = this.client?.create<T>(resource);
+      if (request)
+        request
+          .then((resource) => {
```

Comment:
The `?` in `this.client?.create` is unnecessary because you have a `this.client` check right above it. The `if (request)` is unnecessary too. This could be simplified to:
```suggestion
      if (!this.client) {
        subscriber.error('FhirClientService has not been initialized.');
        return;
      }
      this.client.create<T>(resource).then((resource) => {
```